torch/lib/c10d/TCPStore.hpp

```diff
@@ -14,17 +14,42 @@
 
 namespace c10d {
 
-class TCPStoreDaemon {
+enum class WatchResponseType : uint8_t {
+  KEY_UPDATED,
+  KEY_CREATED,
+  KEY_DELETED
+};
+
+// Abstract base class to handle thread state
+class BackgroundThread {
  public:
-  explicit TCPStoreDaemon(int storeListenSocket);
-  ~TCPStoreDaemon();
+  explicit BackgroundThread(int storeListenSocket);
+  virtual ~BackgroundThread() = 0;
```

Comment:
I am a little confused. This is a pure virtual function, but we still provided implementation for it? C++ reference mentioned that:

> C++ does not prohibit Base from providing a definition for the pure virtual function. However, the derived class still has to implement the pure virtual function. Definition of the pure virtual function has to be placed outside the declaration (in header or source file - doesn't matter).

https://en.cppreference.com/book/intro/abstract_classes

Reply:
Yep, the references seems good, a pure virtual function only **requires** derived classes to override the method, which is orthogonal to whether it is implemented by the base class. The pure virtual destructor prevents the class from being instantiated and the derived classes implicitly override it with their default destructors. The destructors still gets called in the normal order, ~derived() -> ~base()

I just wanted to make `BackgroundThread` an abstract class since C++ doesn't have the `abstract` keyword, if there are any other ways we usually do it let me know!

Reply:
Trying to understand this :)

I didn't see a destructor in `TCPStoreMasterDaemon`, and there is no problem for creating `TCPStoreMasterDaemon` objects. Is it because C++ considers the implicit destructor of `TCPStoreMasterDaemon` overrides `BackgroundThread`'s pure virtual destructor?

Reply:
Yes, exactly!